Rose Bush utility migration to cylc: incoming

sadielbartholomew opened the pull request:
Migration of the 'Rose Bush' HTTP-based service. Closes #1052.

This PR consists of purely the migration to a functional Cylc facility & related 'tidying up' i.e. removing code needed for Rose as a distinct system to retrieve required Cylc data. Follow-up work such as updating the docs & upgrading bootstrap will be completed separately.

### Note:

1. Other than **three core files**, one under `bin/` & two under `lib/`, & **test files** under `tests/`, new files are all **web-based (HTML, CSS & Javascript) files**, under a new directory ``lib/cylc/cylc-nameless/`` with identical structure to that currently in ``rose/lib/html/`` (with files relating to ``rosie disco`` removed & only small modifications made otherwise). These are mostly external. Note that ``jquery.dataTables.js`` constitutes ~3/4 of the ~20,000 line additions for the PR.
2. Commits ``682b311``, ``edd18fc`` & ``b15b25b`` contain unedited copies of Rose files, to allow relative changes to be seen for aid of review.

### Checklist:

- [x] **Agree on name** *--> ``cylc review`` agreed in UM workshop week discussions!*
- [x] **Apply agreed name**.
- [x] Decide on **favicon**: initially used a ``.png`` scaled crop of the cylc logo in ``doc/src``, but maybe want something else &/or ``.ico`` format? *--> Addressed in PR sadielbartholomew/cylc#1*
- [x] Decide how to process **Rose log & config files** (where found): currently the code recognises standard Rose file names & displays such files in a 'rose files' drop-down menu.
- [x] Determine best **locations** (for web-based files relative to `lib/` directory & for `CylcNamelessDAO` class).
- [x] **GPL/copyright**: check (MO & NIWA) legal advice RE code 'donation' Rose -> Cylc: see  #2730, copyright notices to find & replace.
- [x] **Fix tests**.
- [x] **Resolve conflicts**
- [x] **Port security PR** (metomi/rose#2167).
- [x] Amend existing text problem (see comment).

**Open for review & responding to feedback as it is provided**.

oliver-sanders commented:
Travis is none too happy, this might be the problem:

```
NameError: global name 'pre_select_broadcast_states' is not defined
```

sadielbartholomew commented:
I will start working through the issues picked up Travis CI etc when I return. Thanks for the heads-up.

hjoliver commented:
A few name ideas:
 * `cylc suite-log` - boring but intuitive?
 * `cylc slog` - short for previous - snappy, but weird and unpleasant?
 * `cylc bush` - to acknowledge its history, but puzzling to future users?
 * `cylc web-log` - for obvious reasons
 * `cylc blog` - short for previous ... but suggests blogging capability? (so, probably not).

hjoliver commented:
+21k lines - that's expanded the code base!

hjoliver commented:
Noticed the CLI `cylc nameless --help` has inherited a few options that it doesn't need (e.g. `--icp`).

hjoliver commented:
* I'll take a look at the icon format
 * regarding Rose config files - can we just provide access to them if found, otherwise not? (or am I missing the point here?)
 * file and class locations seem fine to me at first glance.

oliver-sanders commented:
> A few name ideas

In the, hopefully, not too distant future, we hope to incorporate "cylc nameless" into the "one cylc gui" so it would be good to agree on a naming which is compatible with this...

| System             | Name (actions)   | Name (descriptions) |
| ------------------ | --------- | ------------ |
| *rose edit (possible plugin)* | *Configure* | *Rose Config*  |
| cylc gui           | Control   | Live         |
| cylc nameless      | Review    | History      |

I like the idea of a name which clarifies the functionality of cylc gui and cylc nameless as this is a source of confusion for new users.

oliver-sanders commented:
As regards the favicon png is fine, https://en.wikipedia.org/wiki/Favicon#File_format_support

sadielbartholomew commented:
Thanks for the suggestions & comments so far all :) I'm working on this on & off (& will comment when the tests adaptation is done) & will read any further comments should you think of anything further to add. In response:

**The name:** Of course it needs to be a joint decision so I think the best way to proceed (if you disagree do say) is for the Metomi team to have a discussion, from which I will summarise the main ideas & points made & report back in a comment here. Hopefully we can narrow our ideas down & try to agree as far as possible as a team before discussing by comment. Though I am glad Oliver commented RE compatibility with future plans as I think that is very important! We will have this discussion sometime next week (as early as possible given people's leave).

**Favicon, locations ...**: great, I will leave these be for now.

@hjoliver Thanks for alerting me about the ``--help`` options, I'll sort that out with the tests. And yes, I was simply asking whether the Rose config files should be displayed at all i.e. 'access if found' as you state. I believe Oliver had pointed out that, for example, if 'standard' Rose files are to be recognised and displayed in this way it may require them to be outlined in the cylc documentation. For (similar?) reasons one might have argued not to acknowledge them at all. But it seems we are all happy with the 'access if found' approach (I discussed this in person with Matt & Oliver). A nod to the existing reference to the Rose docs would probably suffice for the point provided.

sadielbartholomew commented:
@hjoliver, regarding the discussion of the name - various leave meant this was delayed, but we had a conversation about it yesterday, from which I will summarise:

* *Your own ideas*: we thought 'suite-log' would be confusing & 'web-log' inappropriate for future plans given the GUI should replace the web aspect, but simply 'log' is good. We weren't keen on the others.
* The *future vision for the GUI* was our main consideration. We envision a set of names for the function elements, all either verbs or nouns providing intuitive descriptions or actions for each. We lean towards action verbs & are thinking along the lines of:

  | Configuration | Graphing | 'Rose Bush'  |  Analytics  |
  | :-------------: |:-------------: | :-----:| :------:|
  | ``configure`` | ``preview`` ('cylc graph') /``control`` ('gcylc') | ``review`` | ``analyze`` |

Let us know your thoughts on the above.

hjoliver commented:
@sadielbartholomew - yes, I like those suggestions.  Not entirely sure of "review" but I haven't thought of anything better.

sadielbartholomew commented:
Thanks @hjoliver, I will bring this up again in our team meeting tomorrow & mention your comment. If we have any more thoughts I will convey them here. There is no particular urgency on the naming decision yet, as I believe we are still waiting on the legal advice & after I fix the tests & any related issues (will get back onto this soon) there is still the review stage which may take a while.

sadielbartholomew commented:
Re-started work on this PR this afternoon.

There are a large number of issues (250+) as highlighted by Codacy, but almost all are from the web-based files & inherited from the original Rose code. Only a small number are in the Python code & I have removed all of the rectifiable Python issues. So (for this PR at least) we should ignore the Codacy failure.

I do obviously however need to resolve the failures picked up by Travis CI / the test battery & will do that next.

oliver-sanders commented:
There is a minor problem with some of the text in the current rose bush it says:
```
Cycles (before, after or patterns): >CYCLE | <CYCLE | GLOB ...
```
When it should be:
```
Cycles (before, after or patterns): <CYCLE | >CYCLE | GLOB ...
```
Could we possibly sneak this one-line change in with this PR :pray:.

sadielbartholomew commented:
Sure, I'll add that to the checklist. Hope to get through all or most of the items on it today :)

sadielbartholomew commented:
:balloon: :confetti_ball: :beers: Put up the bunting @hjoliver @matthewrmshin @oliver-sanders because this is finally **ready for review** :exclamation: See my comment from 8th June RE the Codacy failure; Travis CI & my local test battery pass & no issues now materialise with manual probing of the service on an ad-hoc server.

I'll very kindly leave one of you to make some obligatory 'reviewing review' joke.

I am on leave from 9th-16th, otherwise I can respond to feedback fairly promptly.

hjoliver commented:
@sadielbartholomew - excellent, bunting "put up", this will be a killer enhancement to cylc!  (Bad news is I am also away 9-16, but I'll try to make a start on this anyway...)

sadielbartholomew commented:
No problem @hjoliver - this is not urgent in terms of milestones, I was just very impatient to get it to the 'ready for review' stage because I have been working on it on & off for months (& adapting certain aspects including the tests was 'a learning curve' to say the least). Please don't look at this whilst on your holiday :)

hjoliver commented:
After an initial play with this, it seems to be working well.

One minor problem is that `cylc review --help` doesn't print accurate or complete information (e.g. `[START]` instead of `start` to start the ad hoc server).  It should probably print the info in the comments in `bin/cylc-review`, no?

sadielbartholomew commented:
Well spotted @hjoliver, I will make sure this output is helpful & accurate. I had put in ``[START]`` etc. early on as placeholders to be corrected but obviously neglected to update & check this.

sadielbartholomew commented:
I am now going to amend the current copyright notices on these files in-line with #2730.

sadielbartholomew commented:
As a note for future improvement (not for this PR though :lock: as it is behaviour inherited from Rose Bush): the CLI for managing ad-hoc servers i.e. the code in ``lib/cylc/ws.py`` has issues.

Namely, with start-up a given server, 3 files are created in ``~/.cylc/`` starting ``cylc-review-<host>-<port>`` & ending ``.status``, ``-error.log``& ``-access.log`` & they are not removed when the server is stopped (cleanly or otherwise). Currently the code, when requested, stops or prints the status of the *first* server (by host & port) it finds one such file for by globbing. So if multiple servers have been started, even if all but one have been stopped, ``cylc review stop`` & ``cylc review`` will generally not be applied to the correct server, hence it is not guaranteed that servers can be stopped cleanly via command unless there are no other server files in ``~/.cylc`` (i.e. if there are old ones they must be deleted manually).

Additionally, ``start [PORT]`` is supported as an argument, but not ``stop [PORT]``, so if users run multiple servers they can only stop them in a random order as above (at least via our CLI, they could of course kill them by PID). Users shouldn't need to, & I imagine won't, do so, but it does strike me as incomplete not to support ``stop [PORT]``.

matthewrmshin commented:
A few other issues:
* On taskjobs or cycles page, clicking the Rose Info menu item hangs.
* On CLI server start, after viewing a page that opens any suite DB, I am unable to Ctrl-C (SIGINT) to shutdown the server. An `lsof -p PID` (where PID is the process ID of the `cylc-review` command) appears to tell me that the server is still hanging onto a suite DB. Are we forgetting to close a DB connection somewhere?

sadielbartholomew commented on lib/cylc/rundb.py:
```diff
@@ -486,14 +503,27 @@ def select_broadcast_states(self, callback, id_key=None):
+        for row_idx, row in enumerate(self.connect().execute(stmt, stmt_args)):
+            callback(row_idx, list(row))
+
+    def pre_select_broadcast_events(self, order=None):
```
Good point, though I assume in this & the other case with an identical comment that you mean ``@classmethod``, since class and not instance variables are referenced, & imply decorating as such after changes to attribute referencing (``self.TABLE_BROADCAST_EVENTS`` can & indeed should be ``cls.TABLE_BROADCAST_EVENTS``). Addressing in my next commit.

sadielbartholomew commented on lib/cylc/rundb.py:
```diff
@@ -476,7 +481,19 @@ def _execute_stmt(self, stmt, stmt_args_list):
             raise
 
-    def select_broadcast_states(self, callback, id_key=None):
+    def pre_select_broadcast_states(self, id_key=None, order=None):
```
[see below]

sadielbartholomew replied:
I've had a look & these (this, plus the method against your identical comment) are not static methods, as far as I can see, as they both reference ``self``, though do not modify instances. I could convert all the ``TABLES_ATTRS`` to class attributes & then make these class methods, but I am do not really think that would be appropriate.

sadielbartholomew replied:
> easily be converted

I've just attempted this & while in theory it should be trivial it seems they cannot be (at least by myself). I will leave this for future.

sadielbartholomew replied:
I think I tried to do this on August 9th too. Will note & leave for future.

> I could convert all the TABLES_ATTRS to class attributes & then make these class methods

sadielbartholomew commented on lib/cylc/cylc-review/template/suite-base.html:
```diff
@@ -0,0 +1,169 @@
+        {% endif -%}
+        {% endfor -%}
+        {% endif -%}
+        {% if info -%}
```
This is an [issue inherited](https://github.com/metomi/rose/blob/master/lib/html/template/rose-bush/suite-base.html) from the original Rose Bush; I have scarcely touched the web-based files. I would like to merge the utility, adapted as necessary for migration but otherwise largely 'as-is', before we address the various shortcomings of Rose Bush. Otherwise this PR could be indefinitely extended outside its domain & reviewing will become ever-more difficult.

sadielbartholomew replied:
I am changing this particular minor issue since you have pointed it out, however.

sadielbartholomew replied:
Perhaps not, I will investigate.

sadielbartholomew replied:
It is not relevant to Cylc so I will just remove it.

Whilst amending this I also noticed a bug whereby Rose ``.version`` files were recognised twice, & have also fixed this in my response commit.

sadielbartholomew commented on lib/cylc/rundb.py:
```diff
@@ -1009,3 +1039,566 @@ def upgrade_pickle_to_json(self):
         return self.connect().execute("VACUUM")
+
+
+class CylcReviewDAO(object):
```
I can see the following arguments as to whether this would be better as an individual module:

* Cylc Review is independent from suite runtime scheduling & therefore should really be placed in a separate file to the central suite runtime DAO.
* ``rundb.py``, with ``CylcReviewDAO`` included, is now too long (>1500 lines);

however...

* Both ``CylcSuiteDAO`` & ``CylcReviewDAO`` are database access objects & hence have similar purposes, use similar modules, & contain similar methods & attributes (some/most of which we might eventually want to unify e.g. as shared module variables & functions after a more detailed audit).
* ``CylcReviewDAO`` queries a ``CylcSuiteDAO`` DB, hence is closely tied to that class.

@cylc/core: thoughts, opinions?

matthewrmshin replied:
(There is already a another module that deals with runtime database stuff - `cylc.suite_db_mgr`. Makes sense to separate new class in separate module so a suite run does not have to load `cylc review` stuff and vice versa.)

sadielbartholomew replied:
Fair enough, I will create a new module for it.

sadielbartholomew commented on lib/cylc/cylc-review/template/suites.html:
```diff
@@ -0,0 +1,213 @@
+
+      <th>
+        title
+      </th>
```
I've removed these table columns as they are no longer strictly relevant i.e. not relevant to Cylc, though in future we might want to provide a option designed for Rose users which when enabled would lift them from Rose config & show them.

sadielbartholomew commented:
For the record (anticipating these as feedback), I have noticed several new issues & will be responding to them as soon as I return from leave, namely ``500 Internal Server Error`` in several forms:

* ``OperationalError: no such table: task_states`` for numerous suites on their 'cycles' or 'taskjobs' pages;
*  ``ProgrammingError: SQLite objects created in a thread can only be used in that same thread ... ``, ditto;
* ``OperationalError: unable to open database file`` for the 'broadcast states' page for suites.

matthewrmshin commented:
The only thing left really is that we are no longer displaying information from `rose-suite.info`. Not sure how users will feel about this, but perhaps we can have a link to the file available in the drop down menu with the other rose logs. Alternatively, we can modify `rose suite-run` to copy the content of `rose-suite.info` into the `log/` directory or into one of the rose logs (that is recognised by `cylc review`).

sadielbartholomew commented:
>  I have noticed several new issues & will be responding to them as soon as I return from leave

Sorting these out now, rather late relative to that optimistic plan given the lower priority of this PR.

> OperationalError: unable to open database file for the 'broadcast states' page for suites.

I have observed that this case is in fact a bug in the original Rose Bush, though I will investigate & fix it within this PR.

sadielbartholomew commented:
Should be all good now! Codacy failures are all either in the CSS or JavaScript files, so inherited from Rose Bush as is, or in the new Python files but irrelevant.

sadielbartholomew commented on tests/lib/bash/test_header:
```diff
@@ -523,6 +549,101 @@ set_test_remote() {
+    if port_is_busy "${TEST_CYLC_WS_PORT}"; then
+        ok "${TEST_NAME}"
+        TEST_CYLC_WS_URL="http://${HOSTNAME}:${TEST_CYLC_WS_PORT}/"
+        echo "${TEST_CYLC_WS_URL}" >&2
```
Oops, well-spotted.

sadielbartholomew commented on bin/cylc-review:
```diff
@@ -0,0 +1,25 @@
+#  ['lib/cylc/review.py' module has formal docstring output for '--help' option]
+#-------------------------------------------------------------------------------
+
+exec python -m cylc.review "$@"
```
Good idea. Let's be consistent.

sadielbartholomew commented:
@cylc/core, I realise we are all very busy & this may not be reviewed for a while, but please can everyone **outline any thoughts they may have on the following** so I can otherwise finalise this (associated amendments will be minor, self-contained & put in a separate commit so shouldn't impede reviewing):

### Rosie (suite) information

From Matt's recent comment:

> we are no longer displaying information from rose-suite.info. Not sure how users will feel about this

Recall that in Rose Bush this information, if found, is parsed & displayed:
* under 'task jobs' and 'cycles' list pages from a pop-up from a button ``rose-suite.info`` on the menu bar alongside the ``cylc files`` & ``rose files`` pull-down menus;
* into separate columns for each compulsory key by suite in the table in a user's suites listing page.

Matt & I have discussed this further. There are **various options** to decide from.

#### Illustrative examples:

1. No longer display any info from ``rose-suite.info``.
2. Still display info [:diamond_shape_with_a_dot_inside:] from ``rose-suite.info``, but do not parse, just read in the file to get & print:
a) all information;
b) only the compulsory information i.e. owner, project & title.
3. Still (provide the option to) display info [:diamond_shape_with_a_dot_inside:] from a (parsed) ``rose-suite.info``, in a 'pluggable' way that supports configurable enabling or disabling of the display of the file e.g. at deployment time.
4. Move all Rosie information to Cylc as suite metadata, i.e. to the ``[meta]`` section of the ``suite.rc`` (the nuclear option?)

#### Display option examples:

 :diamond_shape_with_a_dot_inside:: *For the 'still display info' options, how should we display the information*? Notably, we could do this:

A. from a pop-up item next to the drop-down menus for Cylc & Rose files (as in Rose Bush);
B. from within the Rose files drop-down menu;
C. within the table in separate columns (requires at least primitive parsing);
D. within the table in raw format i.e. printing ``key=value`` lines directly.

hjoliver commented:
@sadielbartholomew - I like the "nuclear" option, now that we have a general suite metadata section (and Rose no longer aspires to support other suite engines in addition to cylc) why not use it? (However, I don't have a deep understanding of how this might affect other aspects of the Rose system).

dpmatthews commented:
My preference is to migrate rose bush into cylc with as few changes in functionality as possible. As long as it works sensibly with pure cylc suites then I don't see a problem. Any changes to where this meta-data is defined in the future would need to be considered as part of the rose suite-run migration.

matthewrmshin commented:
My observation:

* It looks like `rose-suite.info` is not totally absent on the other pages. It can just be linked under Rose Files menu and rendered as any other log files?
* For suites list, probably easiest to do a quick `KEY=VALUE` look up for `project` and `title` from the `rose-suite.info` (if it exists). We can enhance this to read from `suite.rc:[meta]` in the future.

sadielbartholomew commented:
(Codacy spamming: deleting messages which concern the web files lifted & shifted from Rose Bush.)

sadielbartholomew commented:
@hjoliver & @dpmatthews: the 'nuclear' option [4] would certainly be ideal in the long-term, but I should stress I would certainly not include as part of this PR, as this should be merged before we make any significant changes along any lines.

I should have phrased my question to emphasise that the input I need is really **what to do regarding ``rose-suite.info`` for this PR*. Sorry for not making that clear & for including option 4 which I guess is not appropriate for this.

I'll wait in case @oliver-sanders & @kinow want to chip in, but it seems to me from the responses that we should make a fairly trivial relevant change namely that suggested by @matthewrmshin & then put up an issue to eventually, or possibly even soon as part of the rose suite-run migration, to move Rosie info into suite metadata.

sadielbartholomew commented:
@hjoliver (sorry for multiple replies) from verbal conversations our team is happy to go with the plan outlined in my previous comment, though with the '*Rosie info --> suite metadata*' translation as a preliminary suggestion for the future, not a concrete decided plan (disclaimer!).

hjoliver commented:
(I'm not expecting @kinow to have an opinion on this - he doesn't have much Rose experience yet; i.e. no need to wait on comments).

kinow commented:
Hmmm, Codacy seems to be analyzing the newly added `.js` and `.css` files. I think these files can be ignored from the analysis for now.

I've just branched out of your branch @sadielbartholomew , and [added](https://github.com/kinow/cylc/commit/d09b4564d160e839fe561b8c1586efa99939ef03) the `.codacy.yml` file with the exclusion for these files.

Once Travis-CI is done, I should be able to confirm in Codacy that my branch was built and these issues were not added in Codacy (i.e. we wouldn't have these 20 something warnings in this pull request or when these files are changed later).

(@hjoliver related to that other issue about ignoring certain things in Codacy, this is one way for doing that; we probably want to ignore test files to speed up the analysis time and remove unnecessary comments, as well as files from other projects/frameworks like JS/CSS files)

sadielbartholomew commented:
Thank you @kinow, that will be very useful. Clearly we will otherwise get spammed by Codacy & it will be hard to keep track of the non-bot comments!

sadielbartholomew commented:
> It looks like rose-suite.info is now totally absent on the other pages. It can just be linked under Rose Files menu and rendered as any other log files?
For suites list, probably easiest to do a quick KEY=VALUE look up for project and title from the rose-suite.info (if it exists). 

Done! Rose suite info implemented as agreed. **This should now be ready for final review**:

* Please raise any issues you may spot with me ASAP so I can get them rectified, but there shouldn't be many, if any.
* Would it help reviewing (which will not be pleasant for this PR) for me to squash my commits down in a certain way? Or just do a final squash pre-merge?

matthewrmshin commented:
New functionality looks OK. I'll take a final look at the latest commit tonight.

oliver-sanders commented on lib/cylc/review.py:
```diff
@@ -0,0 +1,917 @@
+                rosie_info = {}
+                try:
+                    for line in open(rosie_suite_info, 'r').readlines():
+                        rosie_key, rosie_val = line.strip().split("=", 1)
```
Trying this out with some real-world cases I encountered this traceback:

```
Traceback (most recent call last):
  File "/net/home/h06/osanders/cylc-2/lib/cherrypy/_cprequest.py", line 679, in respond
    response.body = self.handler()
  File "/net/home/h06/osanders/cylc-2/lib/cherrypy/lib/encoding.py", line 230, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/net/home/h06/osanders/cylc-2/lib/cherrypy/_cpdispatch.py", line 60, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/net/home/h06/osanders/cylc-2/lib/cylc/review.py", line 471, in suites
    rosie_key, rosie_val = line.strip().split("=", 1)
ValueError: need more than 1 value to unpack
```

To replicate try viewing a suite list for a user who runs Rosie suites.

sadielbartholomew replied:
It worked for my own Rosie suites, but good spot; looking at the code this would only occur for suites with a ``rose-suite.info`` whereby, other than invalid lines, there are non-setting lines. There aren't any sections for ``rose-suite.info``, so for a valid file really just any comment lines, which I didn't anticipate, though if they contain an ``=`` they would erroneously be parsed. I'll adapt the code to catch these cases.

sadielbartholomew commented on lib/cylc/review.py:
```diff
@@ -0,0 +1,917 @@
+        data["broadcast_states"] = (
+            self.suite_dao.get_suite_broadcast_states(user, suite))
+        if not data["broadcast_states"]:
+            raise cherrypy.HTTPError(404)
```
This behaviour, i.e. raising a HTTP 404 error, was added deliberately after discussions (mainly verbal) between @matthewrmshin & I. However, see my comment from August 9th; with the state of the code at the time (i.e. at commit 725f7bc99df6299b9b7e0ff6a7b692601183167b), I was getting:

> OperationalError: unable to open database file for the 'broadcast states' page for suites.

for suites with no broadcasts. At the time it was suggested it was preferable to raise a 404 i.e. 'Not Found' error than that database error, or an empty page without even a blank table (of headings but no entries) as in the inherited (& current) Rose Bush, though please clarify with Matt for the specific reasons as I am afraid I cannot recall them.

I've tested this with the lines removed as you suggest & it gives a blank page as for Rose Bush, but this is also bad, as we should provide feedback to the user that there are no broadcasts via either an empty table or a message e.g. 'No broadcasts for this suite'.

sadielbartholomew replied:
Perhaps best to revert to Rose Bush behaviour for this PR to get it merged ASAP and raise a issue RE the broadcasts page for suites with no broadcasts.

matthewrmshin replied:
Hi @sadielbartholomew there are 2 parts to this problem.

If you are getting `OperationalError: unable to open database file`, raise a 404 if the SQLite file does not exist (or for very old suite - if the `broadcast_states` table does not exist).

(If you don't have permission to view the file, raise a 403.)

But if the SQLite file exists, then you should have an empty `broadcast_states` table. In which case we should display an empty table.

Do we have some other cases in between? And do we understand why?

sadielbartholomew replied:
Thanks for the clarification @matthewrmshin. Perhaps I misunderstood the 'two part' nature of this when I first made edits on this aspect.

I'll amend the implementation according to the above. Then I will test for all the broadcast states cases I can anticipate & think about the domain we have to cover. But off the top of my head I do not think there are any further cases to deal with.

sadielbartholomew commented:
I've addressed feedback that was quick & easy to amend. I'm now on leave until Friday (26th), but upon my return I will get the final points addressed & hopefully we can merge that same day. Please discuss the best approach for:

> In the case there are no broadcasts

in the meantime. Thanks.

sadielbartholomew commented:
I'm reverting the broadcast states & events listing implementation to that of the current Rose Bush, so for the final two remaining comments, all I have done is remove:

https://github.com/cylc/cylc/blob/bc825776b428e630b794558c224c361da83c2852/lib/cylc/review.py#L131-L132

Therefore for suites with no broadcasts there is a blank page with no error or feedback. It is outside the domain of this PR to improve Rose Bush. In a follow-on PR we can update the implementation (empty table not blank page & convert associated ``pre_select ...`` methods from instance to class methods) for broadcasts.

Given urgency on the next version, let's just get this merged. Thanks.

hjoliver commented:
Codacy is flagging a lot of javascript issues here. Under the circumstances (minimal port of existing Rose Bush) can we nail these in later PRs?

hjoliver commented:
(or is it worth a quick blast through all the new issues - many may be obvious or easy to fix?)

oliver-sanders commented:
A lot of the JS issues are due to bundled code anyway, we should probably exclude these files in a `.codacy.yml` file to suppress non-Cylc code complaints (https://support.codacy.com/hc/en-us/articles/115002130625-Codacy-Configuration-File)

kinow commented:
@oliver-sanders work-in-progress from a few days ago here https://github.com/kinow/cylc/commits/silent-codacy-1, experimenting with the settings from your link. Might revisit it later this week to confirm whether it works or not.

One note about that approach, is that it overrides whatever is on the web interface of Codacy. So the yaml file settings would become the de-facto settings for exclusions (not sure about the rest)

oliver-sanders commented:
Related PR for Rose: https://github.com/metomi/rose/pull/2224

sadielbartholomew commented:
> We will need to reference this code to the cylc ACKNOWLEDGEMENT.md file

Thanks @oliver-sanders, good thinking. I've just done that & squashed it into the previous commit. It was a copy & paste apart from reformatting them in line with the Cylc acknowledgement file, adapting the paths to those now in Cylc, & also (note) changing the Bootstrap version link to match that referenced (checking the tweaked one works).

Brace yourselves - Codacy is about to spam us again!

sadielbartholomew commented:
For information, Oliver spotted an issue relating to broadcast events, which he fixed & I have merged in (my bad, this was hard to test as I struggled to find suites with broadcasts). Otherwise, good to go (again)!

oliver-sanders commented:
I feel like there should be some ceremony, :champagne: :arrow_right: :boat:.
